### PR TITLE
Custom heap comparison

### PR DESCRIPTION
**Pull request by akravc**
Adds an option to use a custom EqualityReduction method.
This allows the use of custom heap comparison predicates inside the comparison procedures.
If the custom option is marked, top level declarations such as globals and constants in the merged file are kept separate for both the reference procedure and the other procedure.

### CODE REVIEW

File: Sources/nuget.config
Comment (atomb) [inline Sources/nuget.config:1]: Is this file necessary? It seems to me as though it's specifying what's already the default behavior.

File: Sources/SymDiff/source/BoogieVerify.cs [hunk 1]
```diff
@@ -992,12 +992,20 @@ public static int RunVerificationTask(VerificationTask vt, Program prog, out boo
                         if (Options.PreciseDifferentialInline)
                         {
                             List<Declaration> consts = prog.TopLevelDeclarations.Where(x => x is Constant).ToList();
-                            ProcessCounterexamplesWOSymbolicOut(
-                                SErrors, globals, vt.Eq.LocVars, vtLeftProcImpl, vtRightProcImpl, consts, [SErrors[0].Cex.Model]);
+                            try {
+                                ProcessCounterexamplesWOSymbolicOut(
+                                    SErrors, globals, vt.Eq.LocVars, vtLeftProcImpl, vtRightProcImpl, consts, [SErrors[0].Cex.Model]);
+                            } catch (Exception e) {
+                                Log.Out(Log.Normal, "Error producing a counterexample.");
```
Comment (atomb) [inline Sources/SymDiff/source/BoogieVerify.cs]: I think I'd use `Log.Error` instead of `Log.Normal` here, unless that forces the program to quit.

File: Sources/SymDiff/source/Options.cs [hunk 2]
```diff
@@ -101,6 +101,17 @@ public enum DAC_ENCODING_OPT { DAC_LINEAR, DAC_NORMAL };
 
         #endregion 
 
+        #region Supporting custom heap comparison predicates
+
+        // Is a custom heap comparison method specified?
+        public static bool CustomHeapComparison = false;
```
Comment (atomb) [inline Sources/SymDiff/source/Options.cs:107]: If `GenerateComparisons` were nullable, you wouldn't need this extra field.

File: Sources/SymDiff/source/WholeProgram.cs [hunk 3]
```diff
@@ -1797,7 +1824,10 @@ public static List<EquivalenceResult> AllInOneMainWithResult(string[] args, Conf
             // Strip/Free contracts
             //////////////////////////////////////////////////////////
             // only "assume" are left, assert p --> assert true, others are removed
-             StripContracts(p1, p2); 
+            // Inlining attributes should be left in for custom heap comparison predicates 
+            if (!Options.CustomHeapComparison) {
```
Comment (atomb) [inline Sources/SymDiff/source/WholeProgram.cs:1828]: I think we may still want to strip the rest of the contracts when doing a custom heap comparison. Maybe it's possible to have this flag just affect whether certain attributes are removed?

File: Sources/SymDiff/source/Options.cs (hunk 2 above)
Comment (atomb) [reply Sources/SymDiff/source/Options.cs:107]: Alternatively, we could set it up so that `GenerateComparison` (with a slightly different type) was, by default, a reference to `EqualityReduction`, and was always called. To customize it, we'd just replace that default binding with a custom one. It might take too much refactoring to do that at the moment, but I at least want to record this possibility somewhere.

**Review by atomb — APPROVED** [review]
I left a couple of comments that could suggest changes, if the level of effort is low enough, but they're not strictly necessary. We could merge this as-is.

File: Sources/nuget.config
Comment (akravc) [reply Sources/nuget.config:1]: Seems to have been auto-generated. Not necessary!

**Review by atomb — APPROVED** [review]
Looks good!
